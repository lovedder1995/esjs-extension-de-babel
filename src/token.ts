export enum TokenType {
  Identifier = 'Identifier',
  Keyword = 'Keyword',
  LeftParen = 'LeftParen',
  RightParen = 'RightParen',
  LeftCurly = 'LeftCurly',
  RightCurly = 'RightCurly',
  Dot = 'Dot',
  Semicolon = 'Semicolon',
  StringLiteral = 'StringLiteral',
  SpecialCharacter = 'SpecialCharacter',
  Whitespace = 'Whitespace',
  Comment = 'Comment',
}

export interface BaseToken {
  type: TokenType
}

export interface NamedToken extends BaseToken {
  name: string
}

export interface ValuedToken extends BaseToken {
  value: string
}

export interface ExtendedToken extends NamedToken {
  compiler: Function
}

export type Token = BaseToken | NamedToken | ValuedToken | ExtendedToken

export type SingleCharacterToken = '(' | ')' | '{' | '}' | '.' | ';'

export const token = {
  identifier(name: string): NamedToken {
    return {
      type: TokenType.Identifier,
      name,
    }
  },
  keyword(name: string): NamedToken {
    return {
      type: TokenType.Keyword,
      name,
    }
  },
  leftParen(): ValuedToken {
    return {
      type: TokenType.LeftParen,
      value: '(',
    }
  },
  rightParen(): ValuedToken {
    return {
      type: TokenType.RightParen,
      value: ')',
    }
  },
  leftCurly(): ValuedToken {
    return {
      type: TokenType.LeftCurly,
      value: '{',
    }
  },
  rightCurly(): ValuedToken {
    return {
      type: TokenType.RightCurly,
      value: '}',
    }
  },
  dot(): ValuedToken {
    return {
      type: TokenType.Dot,
      value: '.',
    }
  },
  semicolon(): ValuedToken {
    return {
      type: TokenType.Semicolon,
      value: ';',
    }
  },
  stringLiteral(value: string): ValuedToken {
    return {
      type: TokenType.StringLiteral,
      value,
    }
  },
  specialCharacter(value: string): ValuedToken {
    return {
      type: TokenType.SpecialCharacter,
      value,
    }
  },
  whitespace(value: string): ValuedToken {
    return {
      type: TokenType.Whitespace,
      value,
    }
  },
  comment(value: string): ValuedToken {
    return {
      type: TokenType.Comment,
      value,
    }
  },
}
