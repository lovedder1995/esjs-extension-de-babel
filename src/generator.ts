import type { NamedToken, Token, ValuedToken } from './token'
import { TokenType } from './token'
import { getDictionary } from './keywords'

export function generate(tokens: Token[], reverse = false): string {
  const dictionary: Map<string, string> = getDictionary(reverse)
  let output = ''
  let current = 0

  /**
   * Checks if we've consumed all the tokens
   */
  function isAtEnd() {
    return current >= tokens.length
  }

  /**
   * Returns true if the current token is of the given type
   */
  function check(type: TokenType) {
    if (isAtEnd()) return false
    return peek().type === type
  }

  /**
   * Returns the current token without consuming it
   */
  function peek() {
    return tokens[current]
  }

  /**
   * Consumes and returns the current token
   */
  function advance() {
    if (!isAtEnd()) current++
    return previous()
  }

  /**
   * Returns the previous token
   */
  function previous() {
    return tokens[current - 1]
  }

  /**
   * Given a list of types, check if the current token matches
   * one of the types and if so, consumes it. Otherwise, returns
   * false.
   */
  function match(...types: TokenType[]): boolean {
    for (const type of types) {
      if (check(type)) {
        advance()
        return true
      }
    }
    return false
  }

  /**
   * Consumes and returns the current token if it matches the provided type.
   * Throws a syntax error if the token doesn't match the type.
   */
  function consume(type: TokenType) {
    if (check(type)) return advance()
    throw new SyntaxError(
      `Unexpected token: Expected ${type}, found ${peek().type}`,
    )
  }

  /**
   * Repeatedly calls `consume` until `current` points to a token of the given
   * type. Assumes `consume` updates the current token pointer.
   */
  function consumeUntil(type: TokenType, consume: () => void) {
    while (!isAtEnd() && !match(type)) consume()

    /**
     * Throw a syntax error if we reached the end of input and didn't
     * find the type we were looking for.
     */
    if (isAtEnd() && previous().type !== type)
      throw new SyntaxError(`Unexpected end of file. Expected ${type}`)
  }

  function compileNamedToken(token: NamedToken) {
    return dictionary.get(token.name) || token.name
  }

  function compileValuedToken(token: ValuedToken) {
    return token.value
  }

  function compileStatement() {
    let statement = ''
    const token = peek()

    switch (token.type) {
      case TokenType.Keyword:
      case TokenType.Identifier:
        statement += compileNamedToken(token as NamedToken)
        break
      case TokenType.SpecialCharacter:
      case TokenType.StringLiteral:
      case TokenType.Whitespace:
      case TokenType.Comment:
      case TokenType.LeftParen:
      case TokenType.RightParen:
      case TokenType.LeftCurly:
      case TokenType.RightCurly:
      case TokenType.Dot:
      case TokenType.Semicolon:
        return compileValuedToken(token as ValuedToken)
      default:
        throw new SyntaxError(`Unexpected token: ${token.type}`)
    }

    return statement
  }

  while (!isAtEnd()) {
    output += compileStatement()
    advance()
  }

  return output
}
