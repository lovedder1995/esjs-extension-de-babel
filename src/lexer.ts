import type { SingleCharacterToken, Token } from './token'
import { token } from './token'

export function tokenize(input: string): Token[] {
  let current = 0
  const tokens = []

  function finishIdentifier() {
    let name = ''
    while (
      !isWhitespace(input[current]) &&
      !isSpecialCharacter(input[current]) &&
      !isSingleCharacter(input[current]) &&
      !isTick(input[current]) &&
      !isUndefined(input[current])
    ) {
      name += input[current]
      current++
    }

    return token.keyword(name)
  }

  function finishSpecialCharacter() {
    let chars = ''
    while (isSpecialCharacter(input[current])) {
      chars += input[current]
      current++
    }

    return token.specialCharacter(chars)
  }

  function finishStringLiteral(tick: string) {
    let value = input[current]
    current++

    while (input[current] && input[current] !== tick) {
      value += input[current]
      current++
    }

    if (isTick(input[current])) {
      value += input[current]
      current++
      return token.stringLiteral(value)
    }

    throw new Error('Unterminated string, expected a closing \', " or `')
  }

  function finishWhitespace() {
    let value = ''
    while (isWhitespace(input[current])) {
      value += input[current]
      current++
    }

    return token.whitespace(value)
  }

  function finishComment(value: string) {
    current += value.length
    return token.comment(value)
  }

  while (current < input.length) {
    const currentChar = input[current]

    const { isComment, end } = checkComment(input, current)
    if (isComment) {
      tokens.push(finishComment(input.slice(current, end + 1)))
      continue
    }

    if (isWhitespace(currentChar)) {
      tokens.push(finishWhitespace())
      continue
    }

    if (isTick(currentChar)) {
      tokens.push(finishStringLiteral(currentChar))
    } else if (isSingleCharacter(currentChar)) {
      tokens.push(getCharToken(currentChar))
      current++
    } else if (isSpecialCharacter(currentChar)) {
      tokens.push(finishSpecialCharacter())
    } else if (isAlpha(currentChar)) {
      tokens.push(finishIdentifier())
    } else {
      throw new Error(`Unknown character: ${currentChar}`)
    }
  }

  return tokens
}

function isAlpha(char: string) {
  return /^[\p{L}\p{Nl}$_][\p{L}\p{Nl}\p{Nd}$\u200C\u200D_]*$/u.test(char)
}

function isWhitespace(char: string) {
  return /\s/.test(char)
}

function isTick(char: string) {
  return ['"', "'", '`'].includes(char)
}

function isSpecialCharacter(char: string) {
  return (
    !isWhitespace(char) &&
    !isAlpha(char) &&
    !isTick(char) &&
    !isSingleCharacter(char)
  )
}

function isUndefined(value: any) {
  return typeof value === 'undefined'
}

function checkComment(input: string, current: number) {
  // Comprueba si estamos en un comentario de una línea
  if (input[current] === '/' && input[current + 1] === '/') {
    let end = current
    while (end < input.length && input[end] !== '\n') end++

    return { isComment: true, end }
  }

  // Comprueba si estamos en un comentario de varias líneas
  if (input[current] === '/' && input[current + 1] === '*') {
    let end = current
    while (
      end < input.length &&
      !(input[end] === '*' && input[end + 1] === '/')
    )
      end++

    return { isComment: true, end }
  }

  return { isComment: false, end: current }
}

const knownSingleCharacters = new Map([
  ['(', token.leftParen],
  [')', token.rightParen],
  ['{', token.leftCurly],
  ['}', token.rightCurly],
  ['.', token.dot],
  [';', token.semicolon],
])

function isSingleCharacter(char: string): char is SingleCharacterToken {
  return knownSingleCharacters.has(char as SingleCharacterToken)
}

function getCharToken(char: SingleCharacterToken) {
  const builder = knownSingleCharacters.get(char)
  return builder!()
}
