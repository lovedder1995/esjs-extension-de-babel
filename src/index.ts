import { tokenize } from './lexer'
import { generate } from './generator'

export function compile(code: string, reverse = false): string {
  const tokens = tokenize(code)

  return generate(tokens, reverse)
}
