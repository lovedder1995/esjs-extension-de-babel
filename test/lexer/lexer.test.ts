import { describe, expect, it } from 'vitest'
import { TokenType } from '../../src/token'
import { tokenize } from '../../src/lexer'

describe('lexer', () => {
  it('tokenizes a simple program', () => {
    const input = `
// Funcion principal: if (a > b) { return a; } else { return b; }

/**
* Funcion principal
* @param {string} mensaje
* @returns {string}
*/
funcion principal(mensaje) {
  const hola = 'hola';
  consola.escribir(hola + mensaje);

  retornar mensaje;
}`

    const tokens = tokenize(input)

    expect(tokens).toMatchSnapshot()

    // Ensure it uses all TokenTypes available
    expect(Object.keys(TokenType).length).toBeGreaterThan(0)

    for (const tokenType of Object.values(TokenType)) {
      if (tokenType === TokenType.Identifier) {
        // Identifiers are not tokenized at this stage
        continue
      }

      try {
        expect(tokens.some((token) => token.type === tokenType)).toBe(true)
      } catch (error) {
        console.error(`Missing token type: ${tokenType}`)
        throw error
      }
    }
  })
})
