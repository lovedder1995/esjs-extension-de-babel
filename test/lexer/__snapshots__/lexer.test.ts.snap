// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lexer > tokenizes a simple program 1`] = `
[
  {
    "type": "Whitespace",
    "value": "
",
  },
  {
    "type": "Comment",
    "value": "// Funcion principal: if (a > b) { return a; } else { return b; }
",
  },
  {
    "type": "Whitespace",
    "value": "
",
  },
  {
    "type": "Comment",
    "value": "/**
* Funcion principal
* @param {string} mensaje
* @returns {string}
*",
  },
  {
    "type": "SpecialCharacter",
    "value": "/",
  },
  {
    "type": "Whitespace",
    "value": "
",
  },
  {
    "name": "funcion",
    "type": "Keyword",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "name": "principal",
    "type": "Keyword",
  },
  {
    "type": "LeftParen",
    "value": "(",
  },
  {
    "name": "mensaje",
    "type": "Keyword",
  },
  {
    "type": "RightParen",
    "value": ")",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "type": "LeftCurly",
    "value": "{",
  },
  {
    "type": "Whitespace",
    "value": "
  ",
  },
  {
    "name": "const",
    "type": "Keyword",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "name": "hola",
    "type": "Keyword",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "type": "SpecialCharacter",
    "value": "=",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "type": "StringLiteral",
    "value": "'hola'",
  },
  {
    "type": "Semicolon",
    "value": ";",
  },
  {
    "type": "Whitespace",
    "value": "
  ",
  },
  {
    "name": "consola",
    "type": "Keyword",
  },
  {
    "type": "Dot",
    "value": ".",
  },
  {
    "name": "escribir",
    "type": "Keyword",
  },
  {
    "type": "LeftParen",
    "value": "(",
  },
  {
    "name": "hola",
    "type": "Keyword",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "type": "SpecialCharacter",
    "value": "+",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "name": "mensaje",
    "type": "Keyword",
  },
  {
    "type": "RightParen",
    "value": ")",
  },
  {
    "type": "Semicolon",
    "value": ";",
  },
  {
    "type": "Whitespace",
    "value": "

  ",
  },
  {
    "name": "retornar",
    "type": "Keyword",
  },
  {
    "type": "Whitespace",
    "value": " ",
  },
  {
    "name": "mensaje",
    "type": "Keyword",
  },
  {
    "type": "Semicolon",
    "value": ";",
  },
  {
    "type": "Whitespace",
    "value": "
",
  },
  {
    "type": "RightCurly",
    "value": "}",
  },
]
`;
